funnel/models/commentvote.py

```diff
@@ -111,14 +111,17 @@ def __init__(self, **kwargs):
         super(Comment, self).__init__(**kwargs)
         self.votes = VoteSpace(type=SPACETYPE.COMMENT)
 
+    @state.transition(None, state.DELETED)
+    def soft_delete(self):
+        self.user = None
+        self.message = ''
+
     def delete(self):
         """
         Delete this comment.
         """
         if len(self.children) > 0:
-            self._state = COMMENTSTATUS.DELETED
-            self.user = None
-            self.message = ''
+            self.soft_delete()
         else:
             if self.parent and self.parent.is_deleted:
```

Comment:
The `is_deleted` property should be replaced with `self.parent.state.DELETED`.

Reply:
Also, we could merge `soft_delete` and `delete` into a single transition. It always transitions to `DELETED` state, always erases user and message, and will optionally remove itself from the database (using `parent.children.remove(self)`) if conditions are right.